README
YubiKey NEO CCID Tools
======================

This package hold tools for the YubiKey NEO in CCID mode.

ykneo-ccid-modeswitch
---------------------

Tool to switch mode of the YubiKey NEO.  The YubiKey NEO can be in
either of three modes:

 1) HID device
 2) CCID device
 3) HID/CCID hybrid device

When the YubiKey NEO is in any non-HID mode you can use this tool to
switch it to any other mode.  To switch the YubiKey NEO from HID mode
to the CCID or HID/CCID mode, use the external "ykpersonalize" tool.

Dependencies
------------

PCSCLite is needed: http://pcsclite.alioth.debian.org/

How to get and install it depends on your operating systems, here are
some hints:

  (Debian/Ubuntu) apt-get install libpcsclite-dev

License
-------

The project is copyrighted by Yubico and licensed under a BSD license.
See the file COPYING for exact wording.  For any copyright year range
specified as YYYY-ZZZZ in this package note that the range specifies
every single year in that closed interval.

Building
--------

The build system uses Autoconf, to set up the build system run:

  ./configure

Then build the code and install the binaries:

  make all install

Using
-----

If your YuibKey NEO is in either the CCID or HID/CCID mode you can use
this tool like this to put it in HID mode:

  ykneo-ccid-modeswitch -m0

or to put it in CCID mode:

  ykneo-ccid-modeswitch -m1

or to put it in HID/CCID mode:

  ykneo-ccid-modeswitch -m2

After the tool has completed, you will need to power-cycle the device
by removing it from the USB port and re-insert it.

Development
-----------

To build this project from version controlled sources you need at
least the following tools:

* autoconf
* automake
* help2man
* gengetopt

How to get and install them depends on your operating systems, here
are some hints:

  (Debian/Ubuntu) apt-get install autoconf automake help2man gengetopt

Then bootstrap the package using

  autoreconf -fvi

And then build as usual, using

  ./configure

followed by

  make

That should produce a normal build directory.

To make releases you will need additional tools including:

* git
* gpg
* git2cl

How to get and install them depends on your operating systems, here
are some hints:

  (Debian/Ubuntu) apt-get install git git2cl

Feedback
--------

See the Google Group yubico-devel:
http://groups.google.com/group/yubico-devel


NEWS
ykneo-ccid-tools NEWS -- History of user-visible changes.     -*- outline -*-

* Version 0.0.1 (unreleased)

** Added command line parameters.

** Improved documentation.

* Version 0.0.0 (released 2012-12-20)

** Initial release.


neo_if.h
/*
Copyright (c) 2012 Yubico AB
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

/*****************************************************************************
**																	        **
**		N E O _ I F  -  NEO Javacard applet interface                       **
**																			**
**		Date		/ Rev		/ Sign	/ Remark							**
**		12-10-29	/ 0.0.0		/ J E	/ Main								**
**																			**
*****************************************************************************/

#ifndef	__NEO_IF_H_INCLUDED__
#define	__NEO_IF_H_INCLUDED__

#define NEO_AID     { 0xa0, 0x00, 0x00, 0x05, 0x27, 0x20, 0x01, 0x01 }

#define INS_YK2_REQ             0x01	// General request (cmd in P1)
#define INS_YK2_OTP             0x02	// Generate OTP (slot in P1)
#define INS_YK2_STATUS          0x03	// Read out status record
#define INS_YK2_NDEF            0x04	// Read out NDEF record

#define NEO_CONFIG_1            0x00	// Configuration 1
#define NEO_CONFIG_2            0x01	// Configuration 2

#define FORMAT_ASCII            0x00	// Output format is ascii
#define FORMAT_ASCII_NO_FMT     0x01	// Ascii, no formatting characters
#define FORMAT_SCANCODE         0x02	// Output format is scan codes

#define SW_BUTTON_REQD          0x6985	// ISO7816 Access condition not met

#endif // __NEO_IF_H_INCLUDED__


NEO_Config.c
/*
 * Copyright (c) 2012 Yubico AB
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *     * Redistributions in binary form must reproduce the above
 *       copyright notice, this list of conditions and the following
 *       disclaimer in the documentation and/or other materials provided
 *       with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <unistd.h>

#ifdef __APPLE__
#include <PCSC/wintypes.h>
#else
#include <winscard.h>
#endif

#include "neo_if.h"
#include "ykdef.h"

#include "cmdline.h"

typedef struct
{
  BYTE cla;
  BYTE ins;
  BYTE p1;
  BYTE p2;
} APDU_1;

typedef struct
{
  BYTE cla;
  BYTE ins;
  BYTE p1;
  BYTE p2;
  BYTE lc;
  union
  {
    BYTE data[0x100];
    DEVICE_CONFIG config;
  };
} APDU_2;

static void
dumpHex (const char *descr, BYTE * buf, int bcnt)
{
  int i, j;
  BYTE *p = buf;

  printf ("%s: %d bytes\n", descr, bcnt);

  for (i = 0;; i += 0x10)
    {
      printf ("%04x:", i);
      for (j = 0; j < 0x10; j++)
	{
	  if (j < bcnt)
	    printf ("%c%02x", (j == 8) ? '-' : ' ', *p++);
	  else
	    printf ("   ");
	}
      printf (" <");
      for (j = 0; j < 0x10; j++, buf++)
	{
	  if (j < bcnt)
	    putchar (isprint (*buf) ? *buf : '.');
	  else
	    putchar (' ');
	}
      printf (">\n");
      if (bcnt <= 0x10)
	break;
      bcnt -= 0x10;
    }
}

static int
modeswitch (int verbose, int autocommit, int mode)
{
  LONG rc;
  SCARDCONTEXT hContext;
  SCARDHANDLE hCard;
  DWORD dwReaders, dwActiveProtocol, dwRecvLength;
  char rdrBuf[1024];
  BYTE neoAID[] = NEO_AID;
  BYTE firstSeq;
  union
  {
    APDU_1 apdu1;
    APDU_2 apdu2;
    BYTE buf[1];
  } apdu;
  union
  {
    BYTE buf[0x100];
    STATUS status;
    struct
    {
      STATUS status;
      DEVICE_CONFIG config;
    } select;
  } rAPDU;
  int usb_mode = 0;
  int usb_mode_seen = false;
  int c;
  char commitbuf[256];
  size_t commitlen;

  // Establish context and find reader(s)

  rc = SCardEstablishContext (SCARD_SCOPE_SYSTEM, NULL, NULL, &hContext);
  if (rc != SCARD_S_SUCCESS)
    {
      fprintf (stderr, "error: SCardEstablishContext failed, rc=%08lx\n", rc);
      return 1;
    }

  rc = SCardListReaders (hContext, NULL, NULL, &dwReaders);
  if (rc != SCARD_S_SUCCESS)
    {
      fprintf (stderr, "error: SCardListReaders failed, rc=%08lx\n", rc);
      SCardReleaseContext (hContext);
      return 1;
    }

  if (dwReaders > sizeof (rdrBuf))
    dwReaders = sizeof (rdrBuf);

  rc = SCardListReaders (hContext, NULL, rdrBuf, &dwReaders);
  if (rc != SCARD_S_SUCCESS)
    {
      fprintf (stderr, "error: SCardListReaders failed, rc=%08lx\n", rc);
      SCardReleaseContext (hContext);
      return 1;
    }

  // Connect to card (if any)

  rc = SCardConnect (hContext, rdrBuf, SCARD_SHARE_SHARED,
		     SCARD_PROTOCOL_T1, &hCard, &dwActiveProtocol);
  if (rc != SCARD_S_SUCCESS)
    {
      fprintf (stderr, "error: SCardConnect failed, rc=%08lx\n", rc);
      SCardReleaseContext (hContext);
      return 1;
    }

  // Transmit NEO select command

  apdu.apdu2.cla = 0;
  apdu.apdu2.ins = 0xa4;	// GP select
  apdu.apdu2.p1 = 0x04;
  apdu.apdu2.p2 = 0;
  apdu.apdu2.lc = sizeof (neoAID);
  memcpy (apdu.apdu2.data, neoAID, sizeof (neoAID));

  dwRecvLength = sizeof (rAPDU);
  rc =
    SCardTransmit (hCard, SCARD_PCI_T1, (BYTE *) & apdu, 5 + sizeof (neoAID),
		   NULL, rAPDU.buf, &dwRecvLength);

  if (rc == SCARD_S_SUCCESS)
    {
      dumpHex ("\nSCardTransmit [NEO select aid]", rAPDU.buf, dwRecvLength);

      // Parse STATUS and DEVICE_CONFIG records

      printf ("\nVersion:       %d.%d.%d\n", rAPDU.select.status.versionMajor,
	      rAPDU.select.status.versionMinor,
	      rAPDU.select.status.versionBuild);
      printf ("Seq:           %d\n", rAPDU.select.status.pgmSeq);
      printf ("Mode:          %02x\n", rAPDU.select.config.mode & MODE_MASK);
      printf ("Flags:         %02x\n",
	      rAPDU.select.config.mode & (~MODE_MASK));
      printf ("CR timeout:    %d\n", rAPDU.select.config.crTimeout);
      printf ("Eject time:    %d\n", rAPDU.select.config.autoEjectTime);
      firstSeq = rAPDU.select.status.pgmSeq;
    }
  else
    printf ("SCardTransmit(1) failed\n");

  fprintf (stderr, "\nCommit? (y/n) [n]: ");
  if (autocommit)
    {
      strcpy (commitbuf, "yes");
      puts (commitbuf);
    }
  else
    {
      fgets (commitbuf, sizeof (commitbuf), stdin);
    }
  commitlen = strlen (commitbuf);
  if (commitbuf[commitlen - 1] == '\n')
    commitbuf[commitlen - 1] = '\0';

  // Transmit set config c ommand

  apdu.apdu2.cla = 0;
  apdu.apdu2.ins = INS_YK2_REQ;
  apdu.apdu2.p1 = SLOT_DEVICE_CONFIG;
  apdu.apdu2.p2 = 0;
  apdu.apdu2.lc = sizeof (DEVICE_CONFIG);

  apdu.apdu2.config.mode = mode;
  apdu.apdu2.config.crTimeout = DEFAULT_CHAL_TIMEOUT;
  apdu.apdu2.config.autoEjectTime = 0;

  dwRecvLength = sizeof (rAPDU);
  rc = SCardTransmit (hCard, SCARD_PCI_T1, (BYTE *) & apdu, 5 + apdu.apdu2.lc,
		      NULL, rAPDU.buf, &dwRecvLength);

  if (rc == SCARD_S_SUCCESS)
    {
      dumpHex ("\nSCardTransmit [NEO write config]", rAPDU.buf, dwRecvLength);

      // Parse STATUS record again

      printf ("\nSeq:           %d\n", rAPDU.select.status.pgmSeq);
      printf ("Update %s", (rAPDU.select.status.pgmSeq == (firstSeq + 1)) ?
	      "successful" : "failed");
    }
  else
    printf ("SCardTransmit(2) failed\n");

  // Disconnect and clean up

  SCardDisconnect (hCard, SCARD_LEAVE_CARD);
  SCardReleaseContext (hContext);

  return 0;
}

int
main (int argc, char *argv[])
{
  struct gengetopt_args_info args_info;
  long mode;
  char *endptr;

  if (cmdline_parser (argc, argv, &args_info) != 0)
    return EXIT_FAILURE;

  mode = strtol (args_info.mode_arg, &endptr, 16);
  if (*endptr != '\0'
      || !(mode == 0x00 || mode == 0x01 || mode == 0x02
	   || mode == 0x80 || mode == 0x81 || mode == 0x82))
    {
      fprintf (stderr, "%s: invalid mode: %s\n", argv[0], args_info.mode_arg);
      return EXIT_FAILURE;
    }

  if (modeswitch (args_info.verbose_flag, args_info.yes_flag, mode) != 0)
    return EXIT_FAILURE;

  return EXIT_SUCCESS;
}


ykdef.h
/*
Copyright (c) 2012 Yubico AB
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

/*****************************************************************************************
**																						**
**		Y K D E F  -  Common Yubikey project header										**
**																						**
**		Date		/ Rev		/ Sign	/ Remark										**
**		06-06-03	/ 0.9.0		/ J E	/ Main											**
**		06-08-25	/ 1.0.0		/ J E	/ Rewritten for final spec						**
**		08-06-03	/ 1.3.0		/ J E	/ Added static OTP feature						**
**		09-06-02	/ 2.0.0		/ J E	/ Added version 2 flags                         **
**		09-09-23	/ 2.1.0		/ J E	/ Added version 2.1 flags (OATH-HOTP)           **
**      10-05-01    / 2.2.0     / J E   / Added support for 2.2 extensions + frame      **
**      11-04-15    / 2.3.0     / J E   / Added support for 2.3 extensions              **
**      11-12-05    / 2.4.0     / J E   / Added support for NFC and NDEF                **
**      12-10-28    / 3.0.0     / J E   / NEO changes                                   **
**																						**
*****************************************************************************************/

#ifndef	__YKDEF_H_INCLUDED__
#define	__YKDEF_H_INCLUDED__

// Slot entries

#define	SLOT_CONFIG					1	// First (default / V1) configuration
#define SLOT_CONFIG2                3	// Second (V2) configuration

#define SLOT_UPDATE1                4	// Update slot 1
#define SLOT_UPDATE2                5	// Update slot 2
#define SLOT_SWAP                   6	// Swap slot 1 and 2

#define SLOT_NDEF                   8	// Write NDEF record
#define SLOT_NDEF2                  9	// Write NDEF record for slot 2

#define SLOT_DEVICE_SERIAL          0x10	// Device serial number
#define SLOT_DEVICE_CONFIG          0x11	// Write device configuration record
#define SLOT_SCAN_MAP               0x12	// Write scancode map

#define SLOT_CHAL_OTP1              0x20	// Write 6 byte challenge to slot 1, get Yubico OTP response
#define SLOT_CHAL_OTP2              0x28	// Write 6 byte challenge to slot 2, get Yubico OTP response

#define SLOT_CHAL_HMAC1             0x30	// Write 64 byte challenge to slot 1, get HMAC-SHA1 response
#define SLOT_CHAL_HMAC2             0x38	// Write 64 byte challenge to slot 2, get HMAC-SHA1 response

#define RESP_ITEM_MASK              0x07	// Mask for slice item # in responses

#define RESP_TIMEOUT_WAIT_MASK      0x1f	// Mask to get timeout value
#define RESP_TIMEOUT_WAIT_FLAG      0x20	// Waiting for timeout operation - seconds left in lower 5 bits
#define RESP_PENDING_FLAG           0x40	// Response pending flag
#define SLOT_WRITE_FLAG             0x80	// Write flag - set by app - cleared by device

#define DUMMY_REPORT_WRITE          0x8f	// Write a dummy report to force update or abort

#define SHA1_MAX_BLOCK_SIZE         64	// Max size of input SHA1 block
#define SHA1_DIGEST_SIZE            20	// Size of SHA1 digest = 160 bits

#define SERIAL_NUMBER_SIZE          4	// Size of device serial number

// Frame structure

#define	SLOT_DATA_SIZE				64

typedef struct
{
  unsigned char payload[SLOT_DATA_SIZE];	// Frame payload
  unsigned char slot;		// Slot # field
  unsigned short crc;		// CRC field
  unsigned char filler[3];	// Filler
} YKFRAME;

// Ticket structure

#define	UID_SIZE					6	// Size of secret ID field

typedef struct
{
  unsigned char uid[UID_SIZE];	// Unique (secret) ID
  unsigned short useCtr;	// Use counter (incremented by 1 at first use after power up) + usage flag in msb
  unsigned short tstpl;		// Timestamp incremented by approx 8Hz (low part)
  unsigned char tstph;		// Timestamp (high part)
  unsigned char sessionCtr;	// Number of times used within session. 0 for first use. After it wraps from 0xff to 1
  unsigned short rnd;		// Pseudo-random value
  unsigned short crc;		// CRC16 value of all fields
} TICKET;

// Activation modifier of sessionUse field (bitfields not uses as they are not portable)

#define	TICKET_ACT_HIDRPT		0x8000	// Ticket generated at activation by keyboard (scroll/num/caps)
#define	TICKET_CTR_MASK			0x7fff	// Mask for useCtr value (except HID flag)

// Configuration structure

#define	FIXED_SIZE				16	// Max size of fixed field
#define	KEY_SIZE				16	// Size of AES key
#define KEY_SIZE_OATH           20	// Size of OATH-HOTP key (key field + first 4 of UID field)
#define	ACC_CODE_SIZE			6	// Size of access code to re-program device

typedef struct
{
  unsigned char fixed[FIXED_SIZE];	// Fixed data in binary format
  unsigned char uid[UID_SIZE];	// Fixed UID part of ticket
  unsigned char key[KEY_SIZE];	// AES key
  unsigned char accCode[ACC_CODE_SIZE];	// Access code to re-program device
  unsigned char fixedSize;	// Number of bytes in fixed field (0 if not used)
  unsigned char extFlags;	// Extended flags
  unsigned char tktFlags;	// Ticket configuration flags
  unsigned char cfgFlags;	// General configuration flags
  unsigned char rfu[2];		// Reserved for future use
  unsigned short crc;		// CRC16 value of all fields
} CONFIG;

// Ticket flags

#define	TKTFLAG_TAB_FIRST			0x01	// Send TAB before first part
#define	TKTFLAG_APPEND_TAB1			0x02	// Send TAB after first part
#define	TKTFLAG_APPEND_TAB2			0x04	// Send TAB after second part
#define	TKTFLAG_APPEND_DELAY1		0x08	// Add 0.5s delay after first part
#define	TKTFLAG_APPEND_DELAY2		0x10	// Add 0.5s delay after second part
#define	TKTFLAG_APPEND_CR			0x20	// Append CR as final character
#define	TKTFLAG_PROTECT_CFG2		0x80	// Block update of config 2 unless config 2 is configured and has this bit set

// Configuration flags

#define CFGFLAG_SEND_REF			0x01	// Send reference string (0..F) before data
#define CFGFLAG_PACING_10MS			0x04	// Add 10ms intra-key pacing
#define CFGFLAG_PACING_20MS			0x08	// Add 20ms intra-key pacing
#define CFGFLAG_STATIC_TICKET		0x20	// Static ticket generation

// V1 flags only

#define	CFGFLAG_TICKET_FIRST		0x02	// Send ticket first (default is fixed part)
#define CFGFLAG_ALLOW_HIDTRIG		0x10	// Allow trigger through HID/keyboard

// V2 flags only

#define	CFGFLAG_SHORT_TICKET		0x02	// Send truncated ticket (half length)
#define CFGFLAG_STRONG_PW1          0x10	// Strong password policy flag #1 (mixed case)
#define CFGFLAG_STRONG_PW2          0x40	// Strong password policy flag #2 (subtitute 0..7 to digits)
#define CFGFLAG_MAN_UPDATE          0x80	// Allow manual (local) update of static OTP

// V2.1 flags only

#define	TKTFLAG_OATH_HOTP			0x40	// OATH HOTP mode
#define CFGFLAG_OATH_HOTP8          0x02	// Generate 8 digits HOTP rather than 6 digits
#define CFGFLAG_OATH_FIXED_MODHEX1  0x10	// First byte in fixed part sent as modhex
#define CFGFLAG_OATH_FIXED_MODHEX2  0x40	// First two bytes in fixed part sent as modhex
#define CFGFLAG_OATH_FIXED_MODHEX   0x50	// Fixed part sent as modhex
#define CFGFLAG_OATH_FIXED_MASK     0x50	// Mask to get out fixed flags

// V2.2 flags only

#define TKTFLAG_CHAL_RESP           0x40	// Challenge-response enabled (both must be set)
#define CFGFLAG_CHAL_YUBICO         0x20	// Challenge-response enabled - Yubico OTP mode
#define CFGFLAG_CHAL_MASK           0x22	// Mask to get out challenge type
#define CFGFLAG_CHAL_HMAC           0x22	// Challenge-response enabled - HMAC-SHA1
#define CFGFLAG_HMAC_LT64           0x04	// Set when HMAC message is less than 64 bytes
#define CFGFLAG_CHAL_BTN_TRIG       0x08	// Challenge-response operation requires button press

#define EXTFLAG_SERIAL_BTN_VISIBLE  0x01	// Serial number visible at startup (button press)
#define EXTFLAG_SERIAL_USB_VISIBLE  0x02	// Serial number visible in USB iSerial field
#define EXTFLAG_SERIAL_API_VISIBLE  0x04	// Serial number visible via API call

// V2.3 flags only

#define EXTFLAG_USE_NUMERIC_KEYPAD  0x08	// Use numeric keypad for digits
#define EXTFLAG_FAST_TRIG           0x10	// Use fast trig if only cfg1 set
#define EXTFLAG_ALLOW_UPDATE        0x20	// Allow update of existing configuration (selected flags + access code)
#define EXTFLAG_DORMANT             0x40	// Dormant configuration (can be woken up and flag removed = requires update flag)

// Flags valid for update

#define TKTFLAG_UPDATE_MASK         (TKTFLAG_TAB_FIRST | TKTFLAG_APPEND_TAB1 | TKTFLAG_APPEND_TAB2 | TKTFLAG_APPEND_DELAY1 | TKTFLAG_APPEND_DELAY2 | TKTFLAG_APPEND_CR)
#define CFGFLAG_UPDATE_MASK         (CFGFLAG_PACING_10MS | CFGFLAG_PACING_20MS)
#define EXTFLAG_UPDATE_MASK         (EXTFLAG_SERIAL_BTN_VISIBLE | EXTFLAG_SERIAL_USB_VISIBLE |  EXTFLAG_SERIAL_API_VISIBLE | EXTFLAG_USE_NUMERIC_KEYPAD | EXTFLAG_FAST_TRIG | EXTFLAG_ALLOW_UPDATE | EXTFLAG_DORMANT)

// NDEF structure

#define	NDEF_DATA_SIZE		        54

typedef struct
{
  unsigned char len;		// Payload length
  unsigned char type;		// NDEF type specifier
  unsigned char data[NDEF_DATA_SIZE];	// Payload size
  unsigned char curAccCode[ACC_CODE_SIZE];	// Access code
} YKNDEF;

// Device configuration block (version 3.0)

typedef struct
{
  unsigned char mode;		// Device mode
  unsigned char crTimeout;	// Challenge-response timeout in seconds
  unsigned short autoEjectTime;	// Auto eject time in seconds    
} DEVICE_CONFIG;

#define MODE_OTP                    0x00	// OTP only
#define MODE_CCID                   0x01	// CCID only, no eject
#define MODE_OTP_CCID               0x02	// OTP + CCID composite
#define MODE_MASK                   0x03	// Mask for mode bits

#define MODE_FLAG_EJECT             0x80	// CCID device supports eject (CCID) / OTP force eject (OTP_CCID)

#define DEFAULT_CHAL_TIMEOUT        15	// Default challenge timeout in seconds

// Scancode mapping (version 3.0)

#define SCAN_MAP    "cbdefghijklnrtuvCBDEFGHIJKLNRTUV0123456789\t\r"
#define SHIFT_FLAG                  0x80	// Flag for shifted scan codes

// Status block

typedef struct
{
  unsigned char versionMajor;	// Firmware version information
  unsigned char versionMinor;
  unsigned char versionBuild;
  unsigned char pgmSeq;		// Programming sequence number. 0 if no valid configuration
  unsigned short touchLevel;	// Level from touch detector
} STATUS;

#define CONFIG1_VALID               0x01	// Bit in touchLevel indicating that configuration 1 is valid (from firmware 2.1)
#define CONFIG2_VALID               0x02	// Bit in touchLevel indicating that configuration 2 is valid (from firmware 2.1)
#define CONFIG1_TOUCH               0x04	// Bit in touchLevel indicating that configuration 1 requires touch (from firmware 3.0)
#define CONFIG2_TOUCH               0x08	// Bit in touchLevel indicating that configuration 2 requires touch (from firmware 3.0)
#define CONFIG_STATUS_MASK          0x0f	// Mask for status bits

// Modified hex string mapping

#define	MODHEX_MAP					"cbdefghijklnrtuv"

// USB vendor ID (VID) and product ID (PID) mapping

#define	YUBICO_VID	                0x1050	// Global vendor ID
#define	YUBIKEY_PID		            0x0010	// Yubikey (version 1 and 2)
#define	NEO_OTP_PID                 0x0110	// Yubikey NEO - OTP only
#define	NEO_OTP_CCID_PID            0x0111	// Yubikey NEO - OTP and CCID
#define	NEO_CCID_PID                0x0112	// Yubikey NEO - CCID only

#endif // __YKDEF_H_INCLUDED__
